### PR TITLE
WIP: Release 5.0

### LINKED ISSUE #39: Posts do not get purged when reverted to draft or pending
This is a bit of a weird edge case. If you have a published post that gets reverted back to `draft` (this would have for `pending` too), it doesn't get purged from Varnish. You have to go and click on the "Empty All" button.

This issue seems to be related to the timing of the call to the `purgePost` method. At that point, WordPress updated the post in the database and it's now a draft. `get_permalink` won't give us a useful link to purge.

I was able to work around the issue using the `pre_post_update` hook. Here's a proof concept solution below:

```php
/**
 * Purges a post from Varnish when their status changes from a publicly accessible link to one without one.
 *
 * This method must be called before WordPress updates the database because VarnishPurger uses functions
 * that use current database values.
 *
 * @param int   $post_id
 * @param array $post_data
 */
function varnish_purge_non_accessible_post($post_id, $post_data) {
    global $purger;

    if (!$purger instanceof VarnishPurger
        || empty($post_data['post_status'])
        || !in_array(get_post_status($post_id), array('publish', 'private', 'trash'))
        || !in_array($post_data['post_status'], array('pending', 'draft'))
    ) {
        return;
    }

    $purger->purgePost($post_id);
}
add_action('pre_post_update', 'varnish_purge_non_accessible_post', 10, 2);
```

I didn't do a PR for this because this function doesn't quite fit in the mould of the `purgePost` and `purgeNoID` methods. You need both the `post_id` and `post_data` to validate the post status before and after the update so that we don't step over what the plugin does already. That said, if this code is adequate, I'm happy to do a PR for it. 😄

### LINKED ISSUE #85: Purging all at customize_save_after
Would it be stupid to do a `varnish_http_purge_events_full` on the `customize_save_after` event ?

### PR DESCRIPTION
**Pull request by Ipstenu**
Jumping over 4.9 (I know) to bring in major changes:

* Remove remote dependancy for scans -- they don't change enough for that to be needed anymore
* Scan now saves last-run data per domain
* Support Health Check (uses saved scan results)
* Multiple IPs (fixes #62 )
* Remove redundant age check (fixes #87)
* Add purge for when someone's using customizer (fixes #85)
* Add purge support for pending and draft statuses (fixes #39)
* Add fix for Cloudflare and sslverify
* Add in DreamPress only schema defaults, which is not my fav thing but...

### CODE REVIEW

File: debugger/cookies.json [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		"type": "bad",
+		"cookie": "trwv.uid",
+		"message": "The plugin Marketo Forms and Tracking is putting down cookies on every page load."
+	},
+	"Optimizely Classic":{
+		"type": "bad",
+		"cookie": "optimizelyBuckets",
+		"message": "The plugin Optimizely Classic is putting down cookies on every page load in order to give people unique experiences to perform it's experiments."
+	},
+	" Optimizely X or Classic":{
```
Comment (0x6d617474) [inline debugger/cookies.json:52]: Extra space seems unintended
```suggestion
	"Optimizely X or Classic":{
```

File: varnish-http-purge.php [hunk 2]
```diff
@@ -35,7 +35,7 @@ class VarnishPurger {
 	 * Version Number
 	 * @var string
 	 */
-	public static $version = '4.8.1';
+	public static $version = '4.9';
```
Comment (0x6d617474) [inline varnish-http-purge.php]: Version mismatch with header
```suggestion
	public static $version = '5.0';
```

File: readme.txt [hunk 3]
```diff
@@ -171,15 +167,15 @@ If your web host set up your service, check their documentation.
 
 = What if I have multiple proxy cache IPs? =
 
-Multiple IPs are not supported at this time.
+You may enter them, separated by a comma, on the settings page.
 
 = What version of Varnish is supported? =
 
-This was built and tested on Varnish 3.x. While it is reported to work on 2.x and 4.x, it is only supported on v3 at this time.
+So far this plugin has been reported to successfully function on Varnish v2 through v5.
```
Comment (0x6d617474) [inline readme.txt]: Anecdotally this works fine with my setup on v6 as well. 
```suggestion
So far this plugin has been reported to successfully function on Varnish v2 through v6.
```

File: changelog.txt [hunk 4]
```diff
@@ -1,5 +1,17 @@
 ## CHANGELOG
```
Comment (0x6d617474) [inline changelog.txt:2]: Sync changelog.txt with changelog from readme.txt
```suggestion
= 5.0 =
* July 2019
* Now purges draft and pending posts (to account for people who un-publish) - props @jerlarke
* Localhost the debugger json. They aren't updated that often, and the remote load is unnecessary.
* More support for Health Check
* Remove strtotime check on Age header - props Matt Fields
* Support for multiple IPs (based on P.Brisson's work)

```

**Review by 0x6d617474 — COMMENTED** [review]
I am very excited for this release! Lots of QoL improvements! :tada: 

I noticed a few things while looking through the changes, and left some minor change suggestions. 

Thanks very much for continuing to work on this!

**Comment by Ipstenu** [comment]
Sorry about the length delay in forward progress. It was held up due to internal work that is directly related to testing on Varnish v 6 :) And yeah, it works 👯‍♀️

**Pull request merged**